aiosmtpd/smtp.py

```diff
@@ -431,34 +476,77 @@ def __init__(
-
-        self._smtp_methods: Dict[str, Any] = {
-            m.replace("smtp_", ""): getattr(self, m)
-            for m in dir(self)
-            if m.startswith("smtp_")
-        }
-
-        self._call_limit_default: int
-        if command_call_limit is None:
-            self._enforce_call_limit = False
-        else:
-            self._enforce_call_limit = True
-            if isinstance(command_call_limit, int):
-                self._call_limit_base = {}
-                self._call_limit_default = command_call_limit
-            elif isinstance(command_call_limit, dict):
-                if not all(map(is_int, command_call_limit.values())):
-                    raise TypeError("All command_call_limit values must be int")
-                self._call_limit_base = command_call_limit
-                self._call_limit_default = command_call_limit.get(
-                    "*", CALL_LIMIT_DEFAULT
+        self._auth_mechs.clear()
+        if self.methods_auth:  # pragma: nobranch
+            log.info(
+                "Available AUTH mechanisms: "
+                + " ".join(
+                    m + "(builtin)" if impl.is_builtin else m
+                    for m, impl in sorted(self.methods_auth.items())
                 )
+            )
+
+    @property
+    def tls_context(self) -> Optional[ssl.SSLContext]:
+        return self._tls_context
+
+    @tls_context.setter
+    def tls_context(self, value: Optional[ssl.SSLContext]):
+        if self._tls_context is not MISSING:
+            if value is None:
+                if self._tls_context is None:
+                    return
+                log.warning("tls_context changed to None")
             else:
-                raise TypeError("command_call_limit must be int or Dict[str, int]")
+                if self._tls_context is None:
+                    log.warning("tls_context is being set")
+                else:
+                    log.warning("tls_context is being replaced")
+        self._tls_context = value
+        if value:
+            problem = {ssl.CERT_NONE, ssl.CERT_OPTIONAL}  # noqa: DUO122
+            if value.verify_mode not in problem:
+                log.warning(
+                    "tls_context.verify_mode not in {CERT_NONE, CERT_OPTIONAL}; "
+                    "this might cause client connection problems"
+                )
+            elif value.check_hostname:
+                log.warning(
+                    "tls_context.check_hostname == True; "
+                    "this might cause client connection problems"
+                )
+
+    @property
+    def require_starttls(self) -> bool:
+        return self._req_starttls and bool(self.tls_context)
+
+    @require_starttls.setter
+    def require_starttls(self, value: bool):
+        self._req_starttls = value
+
+    @property
+    def methods_auth(self) -> Dict[str, _AuthMechAttr]:
+        if not self._auth_mechs:
+            self._auth_mechs = {}
```

Comment:
It is important that we assign a new dict to this attrib, to prevent 'contamination' between SMTP class instantiations.